File: 2022/tests/day15.test.ts
import { readFileSync } from 'fs';
import { part1, part2 } from '../src/day15';

const dayNumber = __filename.match(/day(\d+)\.test\.ts/)?.[1];
const inputFile: string = `input/day${dayNumber}.txt`;

const example: string = `
Sensor at x=2, y=18: closest beacon is at x=-2, y=15
Sensor at x=9, y=16: closest beacon is at x=10, y=16
Sensor at x=13, y=2: closest beacon is at x=15, y=3
Sensor at x=12, y=14: closest beacon is at x=10, y=16
Sensor at x=10, y=20: closest beacon is at x=10, y=16
Sensor at x=14, y=17: closest beacon is at x=10, y=16
Sensor at x=8, y=7: closest beacon is at x=2, y=10
Sensor at x=2, y=0: closest beacon is at x=2, y=10
Sensor at x=0, y=11: closest beacon is at x=2, y=10
Sensor at x=20, y=14: closest beacon is at x=25, y=17
Sensor at x=17, y=20: closest beacon is at x=21, y=22
Sensor at x=16, y=7: closest beacon is at x=15, y=3
Sensor at x=14, y=3: closest beacon is at x=15, y=3
Sensor at x=20, y=1: closest beacon is at x=15, y=3
`.trim();
const input: string = readFileSync(inputFile, 'utf8');

describe('part 1', () => {
  test('example', () => {
    expect(part1(example)).toBe(26);
  });
  test(`solution`, () => {
    expect(part1(input)).toBe(4724228);
  });
});

describe('part 2', () => {
  test('example', () => {
    expect(part2(example)).toBe(56000011);
  });
  test(`solution`, () => {
    expect(part2(input)).toBe(13622251246513);
  });
});

File: 2020/aoc2020/src/test/kotlin/Day15Test.kt
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

@DisplayName("Day 15")
class Day15Test {
    private val example1 = readText("day15_example1.txt")
    private val example2 = readText("day15_example2.txt")
    private val example3 = readText("day15_example3.txt")
    private val example4 = readText("day15_example4.txt")
    private val example5 = readText("day15_example5.txt")
    private val example6 = readText("day15_example6.txt")
    private val example7 = readText("day15_example7.txt")
    private val input = readText("day15.txt")
    private val day = Day15()

    @Nested
    @DisplayName("Part 1")
    inner class Part1 {
        @Test
        fun `Matches example 1`() {
            val answer = day.part1(example1)
            assertEquals(436, answer)
        }

        @Test
        fun `Matches example 2`() {
            val answer = day.part1(example2)
            assertEquals(1, answer)
        }

        @Test
        fun `Matches example 3`() {
            val answer = day.part1(example3)
            assertEquals(10, answer)
        }

        @Test
        fun `Matches example 4`() {
            val answer = day.part1(example4)
            assertEquals(27, answer)
        }


        @Test
        fun `Matches example 5`() {
            val answer = day.part1(example5)
            assertEquals(78, answer)
        }

        @Test
        fun `Matches example 6`() {
            val answer = day.part1(example6)
            assertEquals(438, answer)
        }

        @Test
        fun `Matches example 7`() {
            val answer = day.part1(example7)
            assertEquals(1836, answer)
        }

        @Test
        fun `Matches solution`() {
            val answer = day.part1(input)
            assertEquals(1294, answer)
        }
    }

    @Nested
    @DisplayName("Part 2")
    inner class Part2 {
        @Test
        fun `Matches example 1`() {
            val answer = day.part2(example1)
            assertEquals(175594, answer)
        }

        @Test
        fun `Matches example 2`() {
            val answer = day.part2(example2)
            assertEquals(2578, answer)
        }

        @Test
        fun `Matches example 3`() {
            val answer = day.part2(example3)
            assertEquals(3_544_142, answer)
        }

        @Test
        fun `Matches example 4`() {
            val answer = day.part2(example4)
            assertEquals(261214, answer)
        }

        @Test
        fun `Matches example 5`() {
            val answer = day.part2(example5)
            assertEquals(6_895_259, answer)
        }

        @Test
        fun `Matches example 6`() {
            val answer = day.part2(example6)
            assertEquals(18, answer)
        }

        @Test
        fun `Matches example 7`() {
            val answer = day.part2(example7)
            assertEquals(362, answer)
        }

        @Test
        fun `Matches solution`() {
            val answer = day.part2(input)
            assertEquals(573522, answer)
        }
    }
}
